pennylane_lightning/src/StateVector.hpp

```diff
@@ -15,19 +15,764 @@
  * @file
  * Defines the class representation for quantum state vectors.
  */
+
 #pragma once
 
-#include "typedefs.hpp"
+// Required for compilation with MSVC
+#define _USE_MATH_DEFINES
+
+#include <cmath>
+#include <complex>
+#include <functional>
+#include <set>
+#include <stdexcept>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "Util.hpp"
+
+namespace {
+using namespace std::placeholders;
+using std::bind;
+using std::size_t;
+using std::string;
+using std::vector;
+}; // namespace
 
 namespace Pennylane {
 
-class StateVector {
+template <class fp_t = double> class StateVector {
+  private:
+    using CFP_t = std::complex<fp_t>;
+
+    using Func =
+        std::function<void(const vector<size_t> &, const vector<size_t> &, bool,
+                           const vector<fp_t> &)>;
+
+    using FMap = std::unordered_map<string, Func>;
+
+    static constexpr CFP_t ONE = {1, 0};
+    static constexpr CFP_t ZERO = {0, 0};
+    static constexpr CFP_t IMAG = {0, 1};
+    inline static const fp_t SQRT2 = {static_cast<fp_t>(std::sqrt(2))};
+    inline static const fp_t INVSQRT2 = {
+        static_cast<fp_t>(1 / std::sqrt(2)),
+    };
+
+    const FMap gates_;
+    const std::unordered_map<string, size_t> gate_wires_;
+
+    CFP_t *const arr_;
```

Comment:
This is a carry-over of the original design which does state-prep in numpy. To access the numpy data we point the above pointer to it via the available binding logic. I wish to completely remove the numpy prep (to be followed in a subsequent PR) and replace it with a std::vector in Cpp.

For now, I'd rather not tamper with the existing implementation (shared would be best, but not sure how numpy would handle it), since CPP doesn't own the pointer, merely just operates on it.

Reply:
Sounds like a good plan